zilliqa/src/consensus.rs

```diff
@@ -2146,6 +2126,10 @@ impl Consensus {
             ));
         }
 
+        // Tell the block store to request more blocks if it can.
+        let random_peer = self.get_random_other_peer();
+        self.block_store.request_missing_blocks(random_peer)?;
+
```

Comment:
Might be worth flagging this in an issue/in the docs - if you end up choosing your peer badly (and some folks run peers that refuse to respond to the request_missing_blocks() request), the network could presumably slow down quite significantly?

Also, reading this, it looks as though we will request_missing_blocks() on every block execution, which means that we might not slow the network down, but if anyone is slow about sending us blocks, we will issue multiple overlapping requests for a large number of blocks, which will presumably become a bandwidth issue - particularly if lots of nodes join at the same time?

Reply:
Good questions.

The random peer thing seems problematic to me for a few reasons:
* We don't retry requests for blocks, so if enough nodes doesn't respond to us we will get stuck - #1096 
* We only consider peers who are in the consensus committee which seems overly restrictive - #1097

And I agree that bandwidth could be an issue if many nodes are syncing at once, but I'm not sure how to solve it... My best suggestions would just be making sure the code path for "send blocks in response to a block request" is fast and doesn't contend with other operations, plus rate limiting/scoring for peers to lessen the impact of malicious requests?

Reply:
It is a hard thing to do - I'd:

 * Don't request blocks directly. Ask a strategy function what to do and have it request blocks on your behalf.
 * Write the strategy function so that it requests block ranges from "the best" current peer.
 * The strategy function maintains a set of ranges which we don't have, but want. If you can, separate these into "blocking - need these now" and "just want these sometime".
 *  Strategy function maintains a window and a latency figure for each peer
 * When a block comes in, update the bandwidth and latency for the peer - maybe start with AIMD + fast start.
 * On each tick (fsvo "tick"), fill the windows for random peers .
 * If you _really_ feel bad about it, introduce a NAK. When you have too many requests pending, send some NAKs to close your window to those who are being greedy.
 * That's it.

This is basically TCP, but for blocks. It has the tidy advantage that you incidentally get a map you can display of with whom you are communicating and how well you think they are doing. It has the disadvantage that I just thought of it and it might be utter nonsense :-p

Reply:
Will add this to #1097 and continue discussing there, I agree broadly :) Thanks!